llama-index-integrations/vector_stores/llama-index-vector-stores-vespa/llama_index/vector_stores/vespa/base.py

```diff
@@ -0,0 +1,496 @@
+            ids.append(node.node_id)
+
+        self.app.feed_iterable(
+            data_to_insert,
+            schema=schema or self.default_schema_name,
+            namespace=self.namespace,
+            operation_type="feed",
+            callback=callback,
+        )
+        return ids
+
+    async def async_add(
+        self,
+        nodes: List[BaseNode],
+        schema: Optional[str] = None,
+        callback: Optional[Callable[[VespaResponse, str], None]] = callback,
+        max_connections: int = 10,
+        num_concurrent_requests: int = 1000,
+        total_timeout: int = 60,
+        **kwargs: Any,
+    ) -> List[str]:
+        """
+        Add nodes to vector store asynchronously.
+
+        Args:
+            nodes (List[BaseNode]): List of nodes to add
+            schema (Optional[str]): Schema name in Vespa application to add nodes to. Defaults to `default_schema_name`.
+            max_connections (int): Maximum number of connections to Vespa application
+            num_concurrent_requests (int): Maximum number of concurrent requests
+            total_timeout (int): Total timeout for all requests
+            kwargs (Any): Additional kwargs for Vespa application
+        """
+        semaphore = asyncio.Semaphore(max_concurrent_requests)
+        ids = []
+        data_to_insert = []
+        for node in nodes:
+            metadata = node_to_metadata_dict(
+                node, remove_text=False, flat_metadata=self.flat_metadata
+            )
+            logger.debug(f"Metadata: {metadata}")
+            entry = {
+                "id": node.node_id,
+                "fields": {
+                    "id": node.node_id,
+                    "text": node.get_content(metadata_mode=MetadataMode.NONE) or "",
+                    "metadata": json.dumps(metadata),
+                },
+            }
+            if self.embeddings_outside_vespa:
+                entry["fields"]["embedding"] = node.get_embedding()
+            data_to_insert.append(entry)
+            ids.append(node.node_id)
+
+        async with self.app.asyncio(
+            connections=max_connections, total_timeout=total_timeout
+        ) as async_app:
+            for doc in data_to_insert:
+                async with semaphore:
+                    task = asyncio.create_task(
+                        async_app.feed_data_point(
+                            data_id=doc["id"],
+                            fields=doc["fields"],
+                            schema=schema or self.default_schema_name,
+                            namespace=self.namespace,
+                            timeout=10,
+                        )
+                    )
+                    tasks.append(task)
+
+            results = await asyncio.wait(tasks, return_when=asyncio.ALL_COMPLETED)
+            for result in results:
+                if result.exception():
+                    raise result.exception
+        return ids
+
+    def delete(
```

Comment:
One more comment on delete -- its meant to be a filtered delete, deleting any node that has `ref_doc_id == ref_doc_id` in the metadata -- is this possible with vespa?

Reply:
Yeah, so if you look in the `add`-method, you`ll see that the `node_id` (which is the same as the `rec_doc_id`) is added as Vespa internal doc-id, so this will delete only those. 
Metadata filtering is also possible, but will require a more complex Schema-definition. 

Adding more templates and examples for this will be on the list for next iterations.